### DISCUSSION TITLE
Background task via `asyncio.ensure_future()`?

### DISCUSSION THREAD

**Discussion by ebanner** [fastapi/fastapi#8194] [Questions] [answered]
**Description**

I am kicking off an asynchronous task in the middle of an endpoint. The behavior I expect is to receive the response immediately while the coroutine continues runs in the background. Indeed, this is what happens in a starlette app.

```python
from starlette.applications import Starlette

import asyncio
import time

app = Starlette()

async def sleep_for_10_seconds():
    start = time.time()
    while True:
        now = time.time()
        if now - start > 10:
            break
        asyncio.sleep(1)

@app.route('/')
async def homepage(request):
    asyncio.ensure_future(sleep_for_10_seconds())
    return {'hello': 'world'}

if __name__ == '__main__':
    import uvicorn
    uvicorn.run(app, host='0.0.0.0', port=8000)
```

The `/` endpoint returns immediately.

However, fastapi blocks until the future completes. That is, fastapi takes a full ten seconds to complete the request.

I understand that background tasks are possible via [`fastapi.BackgroundTasks`](https://fastapi.tiangolo.com/tutorial/background-tasks/). **UPDATE: I can't get it working with `fastapi.BackgroundTasks` either #404**. However, I was wondering if it was possible to use just `asyncio` machinery to execute an asynchronous task that does not block an endpoint.

If not, can you point me to the part of the code in fastapi that changes this behavior from starlette? I'd like to understand where this is happening. And also, the reason behind changing this behavior from starlette?

Thanks!

**Additional context**

The background tasks I am interested in doing are IO-bound tasks, not CPU-bound tasks.

**Comment by euri10** [comment, marked as answer]
same reply than for #404 I think :)

**Comment by ebanner** [comment]
My mistake! Thank you for taking the time to address it.

**Comment by tiangolo** [comment]
Thanks for the help here @euri10 and Charles (although I only see Charles comment on the email).

Thanks @ebanner for reporting back and closing the issue.

**Comment by DrPyser** [comment]
@tiangolo I removed my comment when I saw someone else had said the same thing on the linked issue already.

**Comment by DeoLeung** [comment]
since fastapi is using `anyio` now, I wonder what's the best practice now, still using `asynio.ensure_future`?
